src/rust/engine/options/src/config.rs

```diff
@@ -1,18 +1,94 @@
 // Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 use std::fs;
 use std::mem;
 use std::path::Path;
 
+use regex::Regex;
 use toml::value::Table;
 use toml::Value;
 
 use super::id::{NameTransform, OptionId};
 use super::parse::parse_string_list;
 use super::{ListEdit, ListEditAction, OptionsSource};
 
+type InterpolationMap = HashMap<String, String>;
+
+pub(crate) fn interpolate_string(
+    value: String,
+    replacements: &InterpolationMap,
+) -> Result<String, String> {
+    let placeholder_re = Regex::new(r"%\(([a-zA-Z0-9_.]+)\)s").unwrap();
+    let caps_vec: Vec<_> = placeholder_re.captures_iter(&value).collect();
+    if caps_vec.is_empty() {
+        return Ok(value);
+    }
+
+    let mut new_value = String::with_capacity(value.len());
+    let mut last_match = 0;
+    for caps in caps_vec {
+        let m = caps.get(0).unwrap();
+        new_value.push_str(&value[last_match..m.start()]);
+        let placeholder_name = &caps[1];
+        let replacement = replacements.get(placeholder_name).ok_or(format!(
+            "Unknown value for placeholder `{}`",
+            placeholder_name
+        ))?;
+        new_value.push_str(replacement);
+        last_match = m.end();
+    }
+    new_value.push_str(&value[last_match..]);
```

Comment:
This looks conceptually quite similar to https://docs.rs/regex/latest/regex/struct.Regex.html#method.replace_all, which appears to support a closure which could implement the name look-up... except the error handling is a bit awkward. It'd have to be something like:

```rust
let unknown_placeholders = vec![];

let new_value = placeholder_re.replace_all(&values, |caps| {
    let placeholder_name = &caps[0];
    match replacements.get(placeholder_name) {
        Some(replacement) => replacement,
        None => {
            unknown_placeholders.push(placeholder_name);
            ""
        }
    }
})

if !unknown_placeholders.is_empty() {
    return Err(format!(...));
}
```

It seems like it's either awkward error-handling or awkward index-manipulation. What do you think?

Reply:
I looked at this, it's not possible for the reason stated in the docs for that function: 

https://docs.rs/regex/latest/regex/struct.Regex.html#method.replace_all

See the "Fallibility" section. My code here is basically their suggestion for how to deal with fallibility.